src/domain/core/src/services/push_request_planner.rs

```diff
@@ -14,31 +14,53 @@ use opendatafabric::*;
 use thiserror::Error;
 
 use super::sync_service::*;
-use super::RepositoryNotFoundError;
-use crate::{DatasetNotFoundError, GetDatasetError};
+use super::{RemoteTarget, RepositoryNotFoundError};
+use crate::{Dataset, DatasetNotFoundError, GetDatasetError};
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 // Service
 ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
 #[async_trait::async_trait]
-pub trait PushService: Send + Sync {
-    async fn push_multi(
+pub trait PushRequestPlanner: Send + Sync {
+    async fn collect_plan(
         &self,
-        dataset_refs: Vec<DatasetRef>,
-        options: PushMultiOptions,
-        sync_listener: Option<Arc<dyn SyncMultiListener>>,
-    ) -> Vec<PushResponse>;
+        dataset_handles: &[DatasetHandle],
+        push_target: Option<&DatasetPushTarget>,
+    ) -> (Vec<PushItem>, Vec<PushResponse>);
 }
 
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
 #[derive(Debug)]
+pub struct PushItem {
+    pub local_handle: DatasetHandle,
+    pub remote_target: RemoteTarget,
+    pub push_target: Option<DatasetPushTarget>,
+}
+
+impl PushItem {
+    pub fn as_response(
+        &self,
+        result: Result<(SyncResult, Arc<dyn Dataset>), SyncError>,
+    ) -> PushResponse {
+        PushResponse {
+            local_handle: Some(self.local_handle.clone()),
+            target: self.push_target.clone(),
+            result: result.map_err(Into::into),
+        }
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
 pub struct PushResponse {
     /// Local dataset handle, if resolved
     pub local_handle: Option<DatasetHandle>,
     /// Destination reference, if resolved
     pub target: Option<DatasetPushTarget>,
     /// Result of the push operation
-    pub result: Result<SyncResult, PushError>,
+    pub result: Result<(SyncResult, Arc<dyn Dataset>), PushError>,
```

Comment:
maybe extract dataset into a separate field?
```rust
pub target_dataset: Option<Arc<dyn Dataset>>
```

Reply:
Separate field - no, because we have nothing to assign when there is an error.
But, I've extracted a separate structure:

```
pub struct PushResponse {
    /// Local dataset handle, if resolved
    pub local_handle: Option<DatasetHandle>,
    /// Destination reference, if resolved
    pub target: Option<DatasetPushTarget>,
    /// Result of the push operation
    pub result: Result<PushResult, PushError>,
}

pub struct PushResult {
    pub sync_result: SyncResult,
    pub target_dataset: Arc<dyn Dataset>,
}
```